From 8462031e9541a4213e670c88f2d5f1bbcd61e971 Mon Sep 17 00:00:00 2001
From: Yann Gautier <yann.gautier@foss.st.com>
Date: Mon, 2 Sep 2024 11:40:43 +0200
Subject: [PATCH 677/954] fix(stm32mp2): remove mapping of BL2 DT area

To prevent from coding issues that could overwrite DT area, we were
mapping this area as read-only on STM32MP1. But on STM32MP2, we need
this area to put BL31 binary. We were then using dynamic mapping. But
the area is included in the whole SYSRAM memory mapping. This is not
allowed with dynamic mapping. As no other code is running at this step,
and we know what code is running in BL2, just remove this extra
read-only protection for STM32MP2. A message is added after the post
load process of FW-CONFIG file, as BL2 DT area will be overwritten
after that.
This corrects Coverity issue: CID 443168.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 1e7e55a8d5bfee5151dd2c144cce8eadd9ddd24b fix(stm32mp2): remove mapping of BL2 DT area]
Change-Id: Ic01d6a443ecf7721380ef39dc570e2d1627008d0
Signed-off-by: Yann Gautier <yann.gautier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/401414
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 plat/st/stm32mp2/bl2_plat_setup.c | 12 +++++-------
 1 file changed, 5 insertions(+), 7 deletions(-)

diff --git a/plat/st/stm32mp2/bl2_plat_setup.c b/plat/st/stm32mp2/bl2_plat_setup.c
index 16a7734de6..ae1f2c30fe 100644
--- a/plat/st/stm32mp2/bl2_plat_setup.c
+++ b/plat/st/stm32mp2/bl2_plat_setup.c
@@ -381,11 +381,6 @@ void bl2_el3_plat_arch_setup(void)
 
 	configure_mmu();
 
-	/* Prevent corruption of preloaded Device Tree */
-	mmap_add_dynamic_region(DTB_BASE, DTB_BASE,
-				DTB_LIMIT - DTB_BASE,
-				MT_RO_DATA | MT_SECURE);
-
 	if (dt_open_and_check(STM32MP_DTB_BASE) < 0) {
 		panic();
 	}
@@ -706,8 +701,11 @@ int bl2_plat_handle_post_image_load(unsigned int image_id)
 			panic();
 		}
 #endif
-		mmap_remove_dynamic_region(DTB_BASE, DTB_LIMIT - DTB_BASE);
-		inv_dcache_range(DTB_BASE, DTB_LIMIT - DTB_BASE);
+
+		/*
+		 * After this step, the BL2 device tree area will be overwritten
+		 * with BL31 binary, no other data should be read from BL2 DT.
+		 */
 
 		break;
 
-- 
2.53.0

